### ISSUE TITLE
Fehler bei chipTAN (optisch) mit Postbank

### ISSUE THREAD

**Issue by dtvmedia** [nemiah/phpFinTS#269]
Hey, danke erstmal an alle Contributoren für den letzten Release! Ich hab in dem Zuge gesehen das laut den Beispielen jetzt auch chipTAN (optisch) unterstützt wird, was ich direkt mal testen wollte. Allerdings funktioniert das bei mir mit der Postbank nicht. Dabei bekomme ich immer:  `ErrorException: Uninitialized string offset: 0 at /mnt/c/Projekte/app/src/vendor/nemiah/php-fints/lib/Fhp/Model/TanRequestChallengeImage.php:34)`. 
Ich hab die zugehörige Action persistent in der DB gespeichert. Wenn ich die wieder reinlade und `$bin->getData()` anschaue steht da in meinem Fall "0078301100" drin. Da ich mich mit dem Protokoll nicht auskenne durchblicke ich das nicht ganz, aber spätestens das `substr` in Zeile 29 failt, da `$data` dort dann `false `ist in meinem Fall. Irgendwelche Ideen woran es hier scheitert?
Ich nutze btw die aktuelle Version v2.1.

**Comment by Philipp91** [comment]
> `0078301100`

Das klingt nicht nach einem Bild, ist ja viel zu kurz. Was ist denn in `$bin->getData()` vor der Serialisierung drin, also direkt dann wenn du `needsTan()` prüfst? Evtl. kannst du auch die Server-Antwort aufzeichnen und das `HITAN`-Segment hier posten?

Im Code steht: "f.e.: the serialized action was saved as a string, but not base64 encoded". Wenn das ein beliebter Bug ist, vielleicht liegt es ja daran? Viele Systeme beispielsweise brechen Strings bei `\0` (nicht `0`) ab.

[Hier](https://github.com/nemiah/phpFinTS/commit/8663e873d129ad95c4daaab74d81e02709b6ff06) ist ein Patch für mehr Validierung -- aber der kurze String, den du gepostet hast, ist auf jeden Fall invalidate.

**Comment by fbett** [comment]
Wird der persistent String ohne Base64-Codierung irgendwo gespeichert (z.B. Datenbank), so gehen beim anschließenden Lesen bestimmte Zeichen verloren. 
Das muss etwas mit dem ISO-8859-1 Zeichensatz des FinTS-System und der UTF-8 Verwendung innerhalb von PHP bzw. MySQL zu tun haben. 
Ich vermute, dass bestimmte binäre Zeichen als UTF-8 geparst werden.

Die Base64-Codierung ist hier zwingend notwendig.

**Comment by Philipp91** [comment]
Base64 löst sicherlich das Problem auf einfache Weise -- gute Empfehlung :-)

_Notwendig_ ist es hoffentlich nicht, man müsste auch PHP und MySQL dazu bewegen können, binäre Daten (unabhängig von der Kodierung) zu speichern. Auf der MySQL-Seite bräuchte man dazu sicherlich einen der BLOB Typen. PHP-Strings haben sowieso keine Kodierung, man kann sie also für die binären Daten verwenden. Ich speichere bei mir alles auf diese Weise und es funktioniert, wobei ich auch keine photoTAN habe.

**Comment by dtvmedia** [comment]
Danke erstmal für euren Input, ich werde versuchen damit morgen mal weiter zu testen. Aber was ich jetzt noch nicht ganz verstanden hab ist: was muss den Base64 codiert werden? Der komplette serialisierte String oder was meint ihr da? Und wenn die Zeichen erst beim lesen verloren gehen, müsste es in der DB ja noch richtig sein oder?

Momentan mach ich einfach ein `serialize()` auf die Action, speicher die in einem MySQL BLOB und lade das Ganze dann entsprechend wieder und versuche es weiter zu verarbeiten.

**Comment by Philipp91** [comment]
Der Vorschlag ist, stattdessen `base64_encode(serialize(...))` zu machen, und beim Laden umgekehrt.

Am besten wäre es, wenn du das `HITAN`-Segment loggen könntest direkt wenn es vom Server kommt, damit wir sehen was drin ist (ohne Serialisierung und Datenbank und so dazwischen).

**Comment by dtvmedia** [comment]
Ah ok alles klar, Danke 👍 Das werde ich wie gesagt morgen nochmal probieren.

Ich hab aber mal noch aus dem Log von vorhin den Request rausgesucht, müsste der hier sein:

````
[2020-04-19 17:39:12] local.DEBUG: > HNHBK:1:3+000000000145+300+0+1'HKIDN:2:2+280:10010010+9999999999+0+0'HKVVB:3:3+0+0+0+<PRIVATE________________>+1.0'HKTAN:4:6+4+HKIDN'HNHBS:5:1+1'  
[2020-04-19 17:39:12] local.DEBUG: < HNHBK:1:3+000000002771+300+3764763550729000IQxpM2dNE2.A11+1+3764763550729000IQxpM2dNE2.A11:1'HIRMG:2:2+0020::Dialoginitialisierung erfolgreich.+0010::Nachricht entgegengenommen.'HIRMS:3:2:2+0020::Information fehlerfrei entgegengenommen.'HIRMS:4:2:3+1040::BPD nicht mehr aktuell. Aktuelle Version folgt.+1050::UPD nicht mehr aktuell. Aktuelle Version folgt.'HIBPA:5:3:3+14+280:10010010+Postbank Berlin+0+1+300+9999'HIKOM:6:4:3+280:10010010+1+3:https?://hbci.postbank.de/banking/hbci.do::MIM:1'HIPINS:7:1:3+1+1+0+5:50:6:::HKPSA:J:HKCDE:J:HKSAL:N:HKEKA:N:HKTAN:N:HKCDB:N:HKCSB:N:HKCDL:J:HKCCS:J:DKPSA:J:HKCSE:J:HKDSC:J:DKPAE:J:HKCSL:J:HKBME:J:HKKAU:N:HKCMB:N:HKCML:J:HKPRO:N:HKCDN:J:HKDME:J:HKCME:J:HKDMC:J:HKPAE:J:HKSPA:N:HKTAB:N:HKKAZ:N:HKCCM:J'DIPINS:8:1:3+1+1+HKCDE:J:HKSAL:N:HKEKA:N:HKTAN:N:HKCDB:N:HKCSB:N:HKCDL:J:HKCCS:J:DKPSA:J:HKCSE:J:HKDSC:J:DKPAE:J:HKCSL:J:HKBME:J:HKKAU:N:HKCMB:N:HKCML:J:HKPRO:N:HKCDN:J:HKDME:J:HKCME:J:HKDMC:J:HKSPA:N:HKTAB:N:HKKAZ:N:HKCCM:J'HIPAES:9:1:3+1+1+0'DIPAES:10:1:3+1+1'HIPSAS:11:1:3+1+1+0'DIPSAS:12:1:3+1+1'HITANS:13:6:3+1+1+0+N:N:0:910:2:HHD1.3.2OPT:HHDOPT1:1.3.2:chipTAN optisch HHD1.3.2:6:1:Challenge:999:N:1:N:0:2:N:J:00:2:N:9:911:2:HHD1.3.2:HHD:1.3.2:chipTAN manuell HHD1.3.2:6:1:Challenge:999:N:1:N:0:2:N:J:00:2:N:9:912:2:HHD1.4OPT:HHDOPT1:1.4:chipTAN optisch HHD1.4:6:1:Challenge:999:N:1:N:0:2:N:J:00:2:N:9:913:2:HHD1.4:HHD:1.4:chipTAN manuell HHD1.4:6:1:Challenge:999:N:1:N:0:2:N:J:00:2:N:9:920:2:BestSign:BestSign::BestSign:6:2:BestSign:999:N:1:N:0:2:N:J:00:2:N:9:930:2:mobileTAN:mobileTAN::mobileTAN:6:2:mobileTAN:999:N:1:N:0:2:N:J:00:2:N:9'HITABS:14:2:3+1+1+0'HITABS:15:4:3+1+1+0'HIPROS:16:3:3+1+1'HISPAS:17:1:3+1+1+0+J:N:J:urn?:iso?:std?:iso?:20022?:tech?:xsd?:pain.001.003.03:urn?:iso?:std?:iso?:20022?:tech?:xsd?:pain.001.001.03:urn?:iso?:std?:iso?:20022?:tech?:xsd?:pain.008.003.02:urn?:iso?:std?:iso?:20022?:tech?:xsd?:pain.008.001.02'HIKAZS:18:5:3+1+1+90:N:N'HIKAZS:19:6:3+1+1+0+90:N:N'HISALS:20:5:3+1+1'HISALS:21:6:3+1+1+0'HIEKAS:22:3:3+1+1+0+J:N:N:3'HIKAUS:23:1:3+1+1+0'HICCSS:24:1:3+1+1+0'HICSES:25:1:3+1+1+0+0:180'HICSBS:26:1:3+1+1+1+N:J'HICSLS:27:1:3+1+1+1+J'HICDES:28:1:3+1+1+1+4:1:180:00:00:00:12345'HICDBS:29:1:3+1+1+1+N'HICDNS:30:1:3+1+1+1+0:1:180:J:J:J:J:J:J:J:J:J:00:00:00:12345'HICDLS:31:1:3+1+1+1+1:1:N:J'HICCMS:32:1:3+1+1+0+1000:J:J'HICMES:33:1:3+1+1+0+1:180:1000:J:J'HICMBS:34:1:3+1+1+1+N:J'HICMLS:35:1:3+1+1+1'HIDMES:36:1:3+1+1+0+1:30:1:30:1000:J:J'HIBMES:37:1:3+1+1+0+1:30:1:30:1000:J:J'HIDSCS:38:1:3+1+1+1+1:30:1:30::urn?:iso?:std?:iso?:20022?:tech?:xsd?:pain.008.003.02'HIDMCS:39:1:3+1+1+1+1000:J:J:1:30:1:30::urn?:iso?:std?:iso?:20022?:tech?:xsd?:pain.008.003.02'HIUPA:40:4:3+9999999999+1+0'HIUPD:41:6:3+9999999999::280:10010010++9999999999+++anonym'HITAN:42:6:4+4++noref+nochallenge'HNHBS:43:1+1'  
[2020-04-19 17:39:12] local.DEBUG: > HNHBK:1:3+000000000113+300+3764763550729000IQxpM2dNE2.A11+2'HKEND:2:1+3764763550729000IQxpM2dNE2.A11'HNHBS:3:1+2'  
[2020-04-19 17:39:12] local.DEBUG: < HNHBK:1:3+000000000183+300+3764763550729000IQxpM2dNE2.A11+2+3764763550729000IQxpM2dNE2.A11:2'HIRMG:2:2+0010::Nachricht entgegengenommen.'HIRMS:3:2:2+0100::Dialog beendet.'HNHBS:4:1+2'  
[2020-04-19 17:39:12] local.DEBUG: > HNHBK:1:3+000000000423+300+0+1'HNVSK:998:3+PIN:1+998+1+1::0+1:20200419:153912+2:2:13:@8@00000000:5:1+280:10010010:<PRIVATE_______>:V:0:0+0'HNVSD:999:1+@254@HNSHK:2:4+PIN:1+999+1538148+1+1+1::0+1+1:20200419:153912+1:999:1+6:10:19+280:10010010:<PRIVATE_______>:S:0:0'HKIDN:3:2+280:10010010+<PRIVATE_______>+0+1'HKVVB:4:3+14+0+0+<PRIVATE________________>+1.0'HKSYN:5:3+0'HNSHA:6:2+1538148++<PRIVATE_____________>''HNHBS:7:1+1'  
[2020-04-19 17:39:12] local.DEBUG: < HNHBK:1:3+000000000652+300+3764763550955000wkVHJVD7Vh.A11+1+3764763550955000wkVHJVD7Vh.A11:1'HNVSK:998:3+PIN:1+998+1+2::3764763551127000DNM6JVCEVHTYO0+1:20200419:173911+2:2:13:@8@00000000:5:1+280:10010010:<PRIVATE_______>:V:0:0+0'HNVSD:999:1+@392@HIRMG:2:2+0020::Dialoginitialisierung erfolgreich.+3075::Starke Authentifizierung ab dem 2019-09-08 erforderlich.:2019-09-08+3060::Teilweise liegen Warnungen/Hinweise vor.'HIRMS:3:2:3+0020::Information fehlerfrei entgegengenommen.'HIRMS:4:2:4+3920::Meldung unterstützter Ein- und Zwei-Schritt-Verfahren:912:913'HIRMS:5:2:5+0020::Auftrag ausgeführt.'HISYN:6:4:5+3764763551127000DNM6JVCEVHTYO0''HNHBS:7:1+1'  
[2020-04-19 17:39:12] local.DEBUG: > HNHBK:1:3+000000000448+300+3764763550955000wkVHJVD7Vh.A11+2'HNVSK:998:3+PIN:1+998+1+1::3764763551127000DNM6JVCEVHTYO0+1:20200419:153912+2:2:13:@8@00000000:5:1+280:10010010:<PRIVATE_______>:V:0:0+0'HNVSD:999:1+@221@HNSHK:2:4+PIN:1+999+1690771+1+1+1::3764763551127000DNM6JVCEVHTYO0+1+1:20200419:153912+1:999:1+6:10:19+280:10010010:<PRIVATE_______>:S:0:0'HKEND:3:1+3764763550955000wkVHJVD7Vh.A11'HNSHA:4:2+1690771++<PRIVATE_____________>''HNHBS:5:1+2'  
[2020-04-19 17:39:12] local.DEBUG: < HNHBK:1:3+000000000337+300+3764763550955000wkVHJVD7Vh.A11+2+3764763550955000wkVHJVD7Vh.A11:2'HNVSK:998:3+PIN:1+998+1+2::3764763551127000DNM6JVCEVHTYO0+1:20200419:173911+2:2:13:@8@00000000:5:1+280:10010010:<PRIVATE_______>:V:0:0+0'HNVSD:999:1+@78@HIRMG:2:2+0010::Nachricht entgegengenommen.'HIRMS:3:2:3+0100::Dialog beendet.''HNHBS:4:1+2'  
[2020-04-19 17:39:12] local.DEBUG: > HNHBK:1:3+000000000548+300+0+1'HNVSK:998:3+PIN:2+998+1+1::3764763551127000DNM6JVCEVHTYO0+1:20200419:153912+2:2:13:@8@00000000:5:1+280:10010010:<PRIVATE_______>:V:0:0+0'HNVSD:999:1+@350@HNSHK:2:4+PIN:2+912+4638091+1+1+1::3764763551127000DNM6JVCEVHTYO0+1+1:20200419:153912+1:999:1+6:10:19+280:10010010:<PRIVATE_______>:S:0:0'HKIDN:3:2+280:10010010+<PRIVATE_______>+3764763551127000DNM6JVCEVHTYO0+1'HKVVB:4:3+14+0+0+<PRIVATE________________>+1.0'HKTAN:5:6+4+HKIDN+++++++++cT?:Dominique Heinelt_0'HNSHA:6:2+4638091++<PRIVATE_____________>''HNHBS:7:1+1'  
[2020-04-19 17:39:13] local.DEBUG: < HNHBK:1:3+000000000738+300+37647635513420000RxovDSW42.A11+1+37647635513420000RxovDSW42.A11:1'HNVSK:998:3+PIN:1+998+1+2::3764763551127000DNM6JVCEVHTYO0+1:20200419:173911+2:2:13:@8@00000000:5:1+280:10010010:<PRIVATE_______>:V:0:0+0'HNVSD:999:1+@478@HIRMG:2:2+0020::Dialoginitialisierung erfolgreich.+3060::Teilweise liegen Warnungen/Hinweise vor.'HIRMS:3:2:3+0020::Information fehlerfrei entgegengenommen.'HIRMS:4:2:4+3920::Meldung unterstützter Ein- und Zwei-Schritt-Verfahren:912:913'HIRMS:5:2:5+0030::Auftrag empfangen - Sicherheitsfreigabe erforderlich'HITAN:6:6:5+4++TRWU1CED4DSBW20200419173911503+4. Tragen Sie nun die generierte chipTAN für die Karte cT?:Dominique Heinelt_0 ein.+@10@0078301100++cT?:Dominique Heinelt_0''HNHBS:7:1+1'  
````
Danach hab ich das Ganze nur noch gespeichert und es ist beim laden dann entsprechend gefailt.

**Comment by Philipp91** [comment]
Da steht wirklich nur:
```
[challengeHhdUc] => Fhp\Syntax\Bin Object
        (
            [string:protected] => 0078301100
        )
```

Vermutlich muss man das einfach manuell eintippen in das Gerät und die optische Übertragung wird über FinTS nicht untersützt. Laut [Anleitung](https://www.postbank.de/privatkunden/chiptan.html#tab=anleitungen&expand=anleitung-ueberweisung) wäre es auch eine "animierte Grafik", die man ja nicht als PNG speichern kann. Und dass eine FinTS-Endanwendung (sowas wie Hibiscus oder Banking4) GIFs oder ähnliches für animierte Grafiken unterstützt, wäre mir neu, käme mir seltsam vor und würde vmtl. auch nicht stabil funktionieren.

**Comment by dtvmedia** [comment]
Es gibt ja zusätzlich noch chipTAN (manuell), da muss man dann wie du schon sagst den Start Code entsprechend selber eintippen. Aber das ist ja nicht Sinn der Sache bei der optischen Variante, wobei es da gut wäre zu wissen ob der Code dann 1 zu 1 kompatibel zur manuellen Variante wäre.

Aber ich bin mir sicher das FinTS auch die optische Variante an sich unterstützt, da ich das damals mit der Software "Haushaltsbuch7" auch getestet hatte und es da funktionierte. Die "animierte Grafik" ist letztlich ja auch nur ein Code, codiert in diesen weißen, flackernden Strichcodes:
![grafik](https://user-images.githubusercontent.com/18654895/79696815-3c290480-827f-11ea-8361-3e6b0271cfb8.png)
Ich hatte eben mal geschaut wie die Postbank das selber umsetzen. Die haben da quasie wirklich nur einzelne Bilder für die weißen Balken die via JS im richtigen Timing sichtbar geschalten werden, sprich kein GIF oder sowas. Meine Vermutung (ohne jetzt irgendwelche Hintergründe zu kennen) wäre daher das man die Visualisierung aus dem Code selber realisieren müsste.

**Comment by Philipp91** [comment]
Ok :-) Das wäre dann recht aufwändig?

Aber du hast Recht, es ist schon auffällig, dass chipTAN manuell und chipTAN optisch separat angeboten werden. Und wenn es mit der Haushaltsbuch-Software funktioniert, dann warum nicht auch mit phpFinTS.

Also `912: chipTAN optisch HHD1.4` und `913: chipTAN manuell HHD1.4` sind die beiden TAN-Verfahren, die für dich erlaubt sind (laut dem `HIRMS` mit Code `3920`). Und in der Anfrage steht eindeutig `HNSHK:2:4+PIN:2+912`. In der Antwort wiederholt die Bank nicht, welches TAN-Verfahren verwendet wird (weil sie `HNSHK` auslässt, das ist [bekannt](https://github.com/nemiah/phpFinTS/blob/d6f6df70edb23b6704e7e4749ea466ae766bd3cf/lib/Fhp/Protocol/Message.php#L322)). Und dann schickt sie eine Challenge, die irgendwie für das optische Verfahren zu kurz aussieht. Außer es ist vielleicht ein Token oder so, mit dem du anderswo die Challenge abrufen kannst (sowas wie `https://api.postbank.de/challenge/0078301100.gif`)?

Was passiert denn, wenn du so tust, als wäre es eine Challenge für das manuelle Verfahren und die Nummer eintippst? (Aber lieber aufpassen, dass der Zugang nicht gesperrt wird bei zu vielen Versuchen.)

Was passiert, wenn du `913` (also das manuelle Verfahren) wählst?

**Comment by dtvmedia** [comment]
Genau, das müsste dann wahrscheinlich der Flickercode sein, allerdings passt der jetzt nicht wirklich zu dem Aufbau in dem Artikel von dir. Wobei der Code den ich hab, ja auch nur für eine Login TAN ist, da muss ja auch keine IBAN und kein Betrag etc. mit codiert werden. Laut einer anderen Seite gibt da wohl auch noch verschiedene Versionen dieses Codes. 

Jedenfalls gibt es in dem Artikel unten noch eine [React Komponente](https://github.com/basimhennawi/react-chiptan), die ich lokal mal in Plain JS umgeschrieben hab und die funktioniert mit den von denen gegebenen Beispiel Codes ganz gut. Wenn ich aber das Ganze mit dem Code den ich hab probiere, flickert zwar was an sich, aber das Lesegerät meckert dann immer nur rum mit "Fehler 25".
Ich weiß jetzt nicht ob die Library an sich vielleicht nicht ganz up2date ist oder ob der Code noch irgendwie vorverarbeitet/konvertiert werden müsste.

Aber da das wie du schon sagst ja über die Grenzen von phpFinTs hinaus geht, werde ich das wahrscheinlich nicht weiter verfolgen und dann nochmal den manuellen Modus probieren, in der Hoffnung das der dann klappt.

**Comment by Philipp91** [comment]
Ja, `912` und `913` sind beide HHD1.4, die ältere Version ist bei der Postbank zwar noch konfiguriert, wird bei deinem Nutzerkonto aber nicht angeboten.

**Comment by fbett** [comment]
Laut Dokumentation sind mit HHD 1.4 die Codes eher länger geworden, da der Luhn-Algorithmus hinzugefügt wurde. 
Schon für HHD 1.3 kommt mir die Zeichenlänge zu kurz vor.

Und es handelt sich nicht um einen Startcode? Vielleicht mit führenden bzw. abschließenden Nullen?

**Comment by dtvmedia** [comment]
Ich hab eben probiert den String mal als Startcode für die manuelle Eingabe zu nutzen.

- "0078301100": dabei kommt ungültige Eingabe
- "78301100": wird an sich genommen, aber fragt dann Betrag und IBAN ab, was für die Login TAN ja falsch wäre
- "783011": wird ebenfalls genommen, aber nur mit Betrag und IBAN..

Es scheint also kein Startcode zu sein. Wenn ich den manuellen Modus nehme ist der Startcode übrigens "100".

Was mir beim Login über die Postbank selber noch aufgefallen ist, ist das die Flickergrafik beim Login in kürzester Zeit eingescannt werden kann (1sec), während bei den Flickergrafiken vor Überweisungen das Gerät schonmal ganz paar Sekunden braucht um den Code einzuscannen (wegen Betrag und IBAN), was ja darauf deutet das der Login Flickercode entsprechend kürzer wäre oder?

**Comment by fbett** [comment]
Ich habe mal geschaut wie hibiscus dies regelt. 

Dort wird das überlieferte Feld "challenge_hhd_uc" nochmal geparst:
https://github.com/willuhn/hbci4java/blob/f0bf5440dc3d2f7f5c1fc9fbd60a5a019e1f3239/src/org/kapott/hbci/passport/HBCIPassportPinTan.java#L427
https://github.com/willuhn/hbci4java/blob/f0bf5440dc3d2f7f5c1fc9fbd60a5a019e1f3239/src/org/kapott/hbci/manager/FlickerCode.java#L243

**Comment by dtvmedia** [comment]
Hab mal in der Spezifikation etwas rumgeschaut und probiert meinen Code (0078301100) danach aufzuschlüsseln:

- "007": LC
- "83": LS (Länge des Startcodes: 3)
- "01": ControlByte
- "100": Startcode (würde passen da der Startcode in dem Login Fall immer 100 zu sein scheint)

LDE1, LDE2, LDE3 stehen da zwar als fixe Längen drin, aber irgendwo ist auch notiert: "Ist  ein  Datenelement  nicht  belegt,  so  entfällt  jeweils  die  gesamte  Gruppe, bestehend  aus  Länge  und  Wert", demnach wäre das ein gültiger Code und die Mindestlänge für HHD1.4 wäre 7 + Länge Startcode.

**Comment by itoxic97** [comment]
Verstehe ich es jetzt richtig, dass ich mit phpFinTs z.B. Sparda Bank West nicht nutzen kann? Sparda Bank West bietet nur ChipTan via Flickercode an. Ich nutze die browser.php. Gebe es den eine Möglichkeit es umzusetzen? 
Ich hätte gedacht, FinTS liefert einen Code zurück, den lasse ich mit der oben genannten JS in browser.php flickern...

**Comment by dtvmedia** [comment]
Die "chipTAN optisch" Modes kannst du nicht nutzen, da es in diesem Package keine JS Implementierung für Flickercodes gibt. Das hier im Verlauf genannte Package unterstützt keine HHD1.4 Flickercodes (außer deine Bank unterstützt noch ältere Versionen davon). Am besten du prüfst mal was bei dir für konkrete TAN Modes verfügbar sind. Bei meiner Bank (Postbank) ist neben dem optischen Mode noch der manuelle Mode verfügbar der ohne Flickercodes auskommt:

```
    [912] => chipTAN optisch HHD1.4
    [913] => chipTAN manuell HHD1.4
```

**Comment by itoxic97** [comment]
Meine Bank wäre die Sparda Bank West.
Da bei mir ChipTan noch nicht aktiviert ist, weiß ich nicht, was zurückgeliefert wird. Jedoch bekommt die FinTs doch ganz bestimmt etwas von der Bank Schnittstelle geliefert. Kann das nicht dennoch weiterverarbeitet werden, wenn ich beispielsweise die JS dafür baue und den zurückgemeldeten Code flickern lasse.

**Comment by dtvmedia** [comment]
Richtig, zurück kommt so ein Flickercode wie ich in hier im Laufe des Issues mal getestet hatte: 0078301100
Den müsste man theoretisch nur an eine JS Implementierung der Flickercodes geben, die diesen dann rendert und man könnte sein TAN damit generieren. 
Leider gibt es meines Wissens keine aktuelle JS Implementierung auf Github (zumindest hab ich in einer schnellen Suche keine gefunden).

**Comment by itoxic97** [comment]
https://github.com/willuhn/hbci4java/blob/master/src/org/kapott/hbci/manager/FlickerCode.java

Ist das was du meinst? Könnte man die Berechnungen als Vorlage nehmen und einen JS Guru anheuern, um daraus ein Skript zu machen. Würden die Informationen dort reichen oder welche weiteren Informationen werden benötigt?

**Comment by lukas-staab** [comment]
fixed by !357 
sollte das Problem lösen (wobei die Issue vielleicht nicht mehr ganz taufrisch ist...)

**Comment by dtvmedia** [comment]
Danke @lukas-staab! Für die Postbank ist das allerdings (leider) nicht mehr lange relevant, da diese schon seit einiger Zeit ankündigt, dass sie ab Mai 2022 ChipTAN nicht mehr unterstützen werden. Ich hoffe es gibt noch paar andere Banken die ChipTAN noch unterstützen, damit die Arbeit nicht umsonst war.

Der PR wurde eben gemerged, daher schließe ich damit das Issue.
